File: ios/Sources/RisemaxiGraft/Classes/Manifest.swift
import Foundation

/// The signed description of a release. Its raw bytes are what the detached signature covers, so it
/// is decoded only after that signature has been verified.
public struct Manifest: Decodable {
    private static let supportedSchema = 1

    let id: String
    /// Set only on a manifest published to a channel, where it stops a release for one channel being
    /// served on another. The manifest generated for the embedded bundle has no channel to name.
    let channel: String?
    /// The release ordering. Optional on the manifest generated for the embedded bundle: it is only
    /// comparable when the consuming app counts releases and native builds on one scale, and a
    /// consumer that cannot do that omits it rather than supplying a number that does not compare.
    let counter: Int?
    let minNativeBuild: Int
    let notBefore: Int?
    let expiresAt: Int?
    let files: [ManifestFile]
    /// Plugin jsNames the bundle can reach. Absent on a manifest built before contracts existed.
    let requires: [String]

    private enum CodingKeys: String, CodingKey {
        case schema, id, channel, counter, minNativeBuild, notBefore, expiresAt, files, requires
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let schema = try container.decode(Int.self, forKey: .schema)
        guard schema == Manifest.supportedSchema else {
            throw DecodingError.dataCorruptedError(forKey: .schema, in: container, debugDescription: "Unsupported manifest schema: \(schema)")
        }
        id = try container.decode(String.self, forKey: .id)
        channel = try container.decodeIfPresent(String.self, forKey: .channel)
        counter = try container.decodeIfPresent(Int.self, forKey: .counter)
        minNativeBuild = try container.decode(Int.self, forKey: .minNativeBuild)
        notBefore = try container.decodeIfPresent(Int.self, forKey: .notBefore)
        expiresAt = try container.decodeIfPresent(Int.self, forKey: .expiresAt)
        files = try container.decode([ManifestFile].self, forKey: .files)
        requires = try container.decodeIfPresent([String].self, forKey: .requires) ?? []
        guard !files.isEmpty else {
            throw DecodingError.dataCorruptedError(forKey: .files, in: container, debugDescription: "Manifest lists no files.")
        }
    }

    /// The path each digest can be read from in a bundle described by this manifest.
    var hrefBySha256: [String: String] {
        return Dictionary(files.map { ($0.sha256, $0.href) }, uniquingKeysWith: { first, _ in first })
    }
}

File: android/src/main/java/com/risemaxi/graft/classes/Manifest.java
package com.risemaxi.graft.classes;

import androidx.annotation.NonNull;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * The signed description of a release. Its raw bytes are what the detached signature covers, so it
 * is parsed only after that signature has been verified.
 */
public class Manifest {

    private static final int SUPPORTED_SCHEMA = 1;

    @NonNull
    private final String id;

    @NonNull
    private final String channel;

    private final long counter;
    private final long minNativeBuild;
    private final long notBefore;
    private final long expiresAt;

    @NonNull
    private final List<ManifestFile> files = new ArrayList<>();

    public Manifest(@NonNull JSONObject json) throws JSONException {
        int schema = json.getInt("schema");
        if (schema != SUPPORTED_SCHEMA) {
            throw new JSONException("Unsupported manifest schema: " + schema);
        }
        this.id = json.getString("id");
        this.channel = json.getString("channel");
        this.counter = json.getLong("counter");
        this.minNativeBuild = json.getLong("minNativeBuild");
        this.notBefore = json.getLong("notBefore");
        this.expiresAt = json.getLong("expiresAt");
        JSONArray filesJson = json.getJSONArray("files");
        for (int i = 0; i < filesJson.length(); i++) {
            files.add(new ManifestFile(filesJson.getJSONObject(i)));
        }
        if (files.isEmpty()) {
            throw new JSONException("Manifest lists no files.");
        }
    }

    @NonNull
    public String getId() {
        return id;
    }

    @NonNull
    public String getChannel() {
        return channel;
    }

    public long getCounter() {
        return counter;
    }

    public long getMinNativeBuild() {
        return minNativeBuild;
    }

    public long getNotBefore() {
        return notBefore;
    }

    public long getExpiresAt() {
        return expiresAt;
    }

    @NonNull
    public List<ManifestFile> getFiles() {
        return files;
    }

    /**
     * @return The path each digest can be read from in a bundle described by this manifest.
     */
    @NonNull
    public Map<String, String> buildHrefBySha256() {
        Map<String, String> hrefBySha256 = new HashMap<>();
        for (ManifestFile file : files) {
            hrefBySha256.putIfAbsent(file.getSha256(), file.getHref());
        }
        return hrefBySha256;
    }
}
